File: 53.maximum-subarray.py
#
# @lc app=leetcode id=53 lang=python3
#
# [53] Maximum Subarray
#
from typing import *
# @lc code=start
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        max_value = - 1 << 32
        last = - 1 << 32

        for i in range(len(nums)):
            last = (max(last + nums[i],nums[i]))
            max_value = max(max_value,last)
        return max_value      
# @lc code=end

if __name__ == "__main__":
    print(Solution().maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))

File: 53.maximum-subarray.go
/*
 * @lc app=leetcode id=53 lang=golang
 *
 * [53] Maximum Subarray
 */
package main

// @lc code=start
func maxSubArray(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	dp := []int{nums[0]}
	for i := 1; i < len(nums); i++ {
		if dp[i-1] < 0 {
			dp = append(dp, nums[i])
		} else {
			dp = append(dp, dp[i-1]+nums[i])
		}
	}

	max := dp[0]
	for i := 0; i < len(dp); i++ {
		if max < dp[i] {
			max = dp[i]
		}
	}
	return max
}

// @lc code=end
